### DISCUSSION TITLE
apply conversion settings from cropped section of image to whole image

### DISCUSSION THREAD

**Discussion by rabanito75** [ImageMagick/ImageMagick#2168] [Help]
Hello community, i'm in the middle of converting thousands of analog color negatives to digital positives. I convert the camera-captured raw file to a 16-bit-tiff file with dcraw, then use ImageMagick to process the image further. 

I work with version 6.9.1-0 Q16 x86_64 2015-03-22, MacOS High-Sierra

My problem is, that i would like to clip the channels of the actual picture to the black and white points but do the cropping only later in Photoshop. But the surrounding area of the actual picture including the sprocket holes, and parts of the filmholder make this difficult.

I tried following code:

convert scan.tiff -negate -region 4310x2800+160+320 -auto-level -gamma 0.5 +region test.tiff

This leaves me with the channels inside the chosen region perfectly clipped, but leaves everything outside the region unchanged (see attached downsized lowers-picture). Is there a possibility to apply the conversion settings from the cropped region to the whole image?

![scan](https://user-images.githubusercontent.com/67098233/85006087-f6db6300-b159-11ea-8f9f-beac8c038d5c.jpg)

I hope one can understand what i mean, English is not my native language... 

Thanks a lot in advance for any advice!

**Comment by snibgo** [comment]
"-auto-level" stretches or shrinks pixel values so they span values from 0 to QuantumRange. We can find the minimum and maximum values of a crop, and apply that "-level" to the original image.

With IM v7, this is fairly simple. Windows syntax:
```
magick scan.tiff -negate ( +clone -crop 4310x2800+160+320 +repage -set option:mymin %[fx:minima*QuantumRange] -set option:mymax %[fx:maxima*QuantumRange] +delete ) -level %[mymin]x%[mymax] -gamma 0.5 test.tiff
```
For bash, escape the parentheses \( and \).

For IM v6, "-level" can't use fx[...] expressions, so it needs two commands. Bash syntax:
```
declare $(convert \
  scan.tiff -negate \
  -negate \
  -crop 4310x2800+160+320 +repage \
  -format "mymin=%[fx:minima*QuantumRange]\n \
     mymax=%[fx:maxima*QuantumRange]" \
  info:)

echo hello $mymin $mymax

convert \
  scan.tiff -negate \
  -level $myminx$mymax -gamma 0.5 \
  test.tiff
```

**Comment by rabanito75** [comment]
Thank you for your reply! 
The code works, but unfortunately, the resulting picture is blown:

![blown](https://user-images.githubusercontent.com/67098233/85033476-484b1880-b181-11ea-8422-8029cfd07962.jpg)

The evaluated min/max-values are "hello 813 15723"

I tried to swap those values with this command:

convert scan.tiff -level 15723x813 -gamma 0.5 test.tif

And to my surprise, the resulting picture is inverted and clipped to the desired values:

![test](https://user-images.githubusercontent.com/67098233/85034460-6a916600-b182-11ea-8762-d9cb26551ea7.jpg)

However, i can't explain why ;–)... do you have an explanation?

Here is the downsized picture out of dcraw to test:

![scan](https://user-images.githubusercontent.com/67098233/85035056-0cb14e00-b183-11ea-8db3-2bbe8a988a93.jpg)

Thank you very much for your help!

**Comment by snibgo** [comment]
The difference is that your original command, and my command, included "-negate". Your latest command does not. So swapping the values has the same effect as "-negate" with the values as I had them.

Your latest command, with no "-negate", will be faster.

Incidentally, the "-level" values should be separated with a comma not "x". My mistake, sorry, though it makes no difference. See https://www.imagemagick.org/script/command-line-options.php#level
```
-level $mymin,$mymax
```

**Comment by rabanito75** [comment]
Thank you for your answer, sorry for the late reply.

So i managed to get the code to work like this by swapping the min and max, automatically inverting the image and clipping the channels to fill the histogram of the later cropped file, very cool, thank you !!

```
declare $(convert scan.tiff \
-crop 4310x2800+160+320 +repage \
-format "mymin=%[fx:minima*QuantumRange] mymax=%[fx:maxima*QuantumRange]" info:)
convert scan.tiff \
-level $mymax,$mymin -gamma 0.5 \
scan_out.tiff
```

Is there a possibility to use the command in a loop to batch process 36 tiff files of a whole film roll?

I tried:

```
for i in *.tiff; do \
declare $(convert *.tiff \
-crop 4310x2800+160+320 +repage \
-format "mymin=%[fx:minima*QuantumRange] mymax=%[fx:maxima*QuantumRange]" info:)
convert *.tiff \
-level $mymax,$mymin -gamma 0.5 \
"$i"; done
```

but this does not work, because it continuously works on the tiff files again.

How can this loop work correctly?

**Comment by snibgo** [comment]
Within each loop, your variable "i" is the name of a file. It loops through all of the files. I suggest that "i" is a poor name for that variable. You might use "myfile" or something.

But then within your loop, you attempt to process every file with "*.tiff". Instead, you should use your variable.
